[PVR] fix compilation warnings

janbar opened the pull request:
Fix comparing signed and unsigned values in many places

janbar commented:
hmmm.  I don't understand your logic because iClientIndex is unsigned and only provided by addon.

fritsch commented on xbmc/pvr/timers/PVRTimerInfoTag.cpp:
```diff
@@ -43,7 +43,7 @@ CPVRTimerInfoTag::CPVRTimerInfoTag(bool bRadio /* = false */) :
 {
   m_iClientId           = g_PVRClients->GetFirstConnectedClientID();
-  m_iClientIndex        = -1;
+  m_iClientIndex        = static_cast<unsigned int>(-1);
```
erm?

fritsch replied:
https://github.com/janbar/xbmc/blob/fix_compilation_warnings/xbmc/pvr/timers/PVRTimerInfoTag.cpp#L679 <- ?

janbar replied:
Seems author want to set it to max int

fritsch replied:
~0 would do.

I talked with @opdenkamp and one just needs some "default" value ... and zero is a bad idea.

janbar replied:
The PR #7439 changed the type unsigned to signed. Now i get compilation warnings. Here i resolve the issue.

fritsch replied:
that cast is not needed - is all I wanted to say ~0U should do.

janbar replied:
Yes too. I just prefer cast because it is portable.

fritsch replied:
I would do something else, if you were really interested in the resulting value.

janbar replied:
@fritsch , i didn't read all code about but i think this value isn't really important and it is evaluated only in default ctor.

janbar replied:
The addon transfers timer with a known value from itself.

janbar replied:
In my opinion the best initial value would be 0 here, same as iParentClientIndex.

fritsch replied:
So then - go ahead :-)

janbar replied:
Do you want i rebase with value 0 ?

ksooo commented:
This is not "my" logic. This was not introduced by the series recording PR. Signed vs unsigned conflict between Kodi and add-on API was already there the first day I looked at Kodi code base. I appreciate that you are going to fix this, but a simple cast will not do, imo.

janbar commented:
In PR #7439 commit 4bff34761439aac0142b305119aa5821529fa407 has probably a typo

ksooo commented:
#7439 reverted a change I've done in #7079. The latter led to compiler and Coverity warnings, too. Now we have the same situation in master as we had before. Thus, neither changing iclientindex to signed not just casting away signed is the right solution. We need something didn't, I guess.

ksooo commented:
s/didn't/different

janbar commented:
@ksooo, please try to compile with this PR and you will see all warnings will disappear.

janbar commented:
@fritsch do you like it now ?

janbar commented:
If it is okay for you , i will rebase.

janbar commented:
@ksooo do you want keep max int ? Then i truncate last commit.

ksooo commented:
Making m_iClientIndex unsigned is not okay, at least if other code pieces are left unchanged. Example: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/windows/GUIWindowPVRTimers.cpp#L349 Coverity complains here, that an unsigned value never will be less then zero. Also, there are other places where m_iClientIndex gets set to -1. I also don't like misusing max int to indicate "invalid client index".

What we need is a way to indicate that a timer is new, just created on the Kodi-side, not yet passed to the backend and therefore not has a backend supplied index.

Just wondering wheter member m_state of PVRTimerInfoTag already contains the needed information, as PVR_TIMER_STATE_NEW (https://github.com/xbmc/xbmc/blob/master/xbmc/addons/include/xbmc_pvr_types.h#L166) seems to serve the same purpose.

Thus, a solution could be to make m_iClientIndex unsigned, initialize it with zero (where now -1 is used) and to change all places in Kodi checking m_iClientIndex for -1 (<0 whatever) to query instead for m_state == PVR_TIMER_STATE_NEW. Maybe there is just the one place in code I mentioned above. Don't know.

@opdenkamp what do you think?

janbar commented:
@ksooo, no problem to change all these stuff. Currently code compare signed with unsigned value and this PR is there to resolve those points. We could also fix usage of magic numbers as you mentioned above. I think using 0 for a new timer is better avoiding weird stuff.
So i will push a other commit to fix them.

ksooo commented:
@janbar To make it clear. The point is not that with my proposed solution zero means "new timer". Zero is today and can be after the change proposed here be a valid backend supplied index. If one wants to know whether a timer has a valid index he/she must not check m_iClientIndex for zero then, but must check m_state for  PVR_TIMER_STATE_NEW.

janbar commented:
Since always we have checked for PVR_TIMER_STATE_NEW (frodo)

janbar commented:
So i dislike use magic number. What do you prefer for the name of the constant:
PVR_TIMER_NULL_ID ?

ksooo commented:
Zero is not a magic number for m_iClientIndex! Zero is just a valid index, like 1, 2 ... 1000. No need for a special constant here, because 0 does NOT denote any special index.

janbar commented:
Rebased

ksooo commented on xbmc/addons/include/xbmc_pvr_types.h:
```diff
@@ -96,6 +96,7 @@ extern "C" {
    * @brief special PVR_TIMER.iParentClientIndex value to indicate that a timer has no parent.
    */
   const unsigned int PVR_TIMER_NO_PARENT = 0; /*!< @brief timer has no parent; it was not scheduled by a repeating timer. */
+  const unsigned int PVR_INVALID_TIMER_ID = 0; /*!< @brief timer has no defined client id. */
```
No, no and no!!! Sorry, but now I'm getting really exhausted. I told you again and again that 0 is no special value!

ksooo commented on xbmc/pvr/timers/PVRTimerInfoTag.h:
```diff
@@ -233,7 +233,7 @@ namespace PVR
     PVR_TIMER_STATE       m_state;               /*!< @brief the state of this timer */
     int                   m_iClientId;           /*!< @brief ID of the backend */
-    int                   m_iClientIndex;        /*!< @brief index number of the tag, given by the backend, -1 for new */
+    unsigned int          m_iClientIndex;        /*!< @brief index number of the tag, given by the backend, PVR_INVALID_TIMER_ID for new */
```
Remove this PVR_INVALID_blah nonsense, please.

opdenkamp replied:
the thing needs to be initialised with some value, or the compiler will complain. and instead of using a magic value, using a #define or const defined somewhere is the proper way to do it.

but the value should not be used for any checks in kodi, that's correct. it's something that's used by the add-on.

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -521,7 +521,7 @@ bool CPVRTimers::InstantTimer(const CPVRChannelPtr &channel)
     newTimer.reset(new CPVRTimerInfoTag);
     /* set the timer data */
-    newTimer->m_iClientIndex      = -1;
+    newTimer->m_iClientIndex      = PVR_INVALID_TIMER_ID;
```
Giving up...

ksooo replied:
What about the other places where m_iClientIndex is assigned -1. Do a grep, man!

ksooo commented on xbmc/pvr/windows/GUIWindowPVRTimers.cpp:
```diff
@@ -346,7 +346,7 @@ bool CGUIWindowPVRTimers::ActionDeleteTimer(CFileItem *item)
   /* check if the timer tag is valid */
   CPVRTimerInfoTagPtr timerTag = item->GetPVRTimerInfoTag();
-  if (!timerTag || timerTag->m_iClientIndex < 0)
+  if (!timerTag || timerTag->m_state == PVR_TIMER_STATE_NEW)
```
This one is correct.

fritsch commented:
Guys. Now both of you calm down a bit - everybody is drinking a beer and then talking can start again a bit more nicely minded.

janbar commented:
Yes i think too. ;) . It is beer time. We will finish tomorrow. tchin @fritsch .

opdenkamp commented:
`m_iClientIndex` should be unsigned in the code because that's what's defined in https://github.com/xbmc/xbmc/blob/master/xbmc/addons/include/xbmc_pvr_types.h#L375

opdenkamp commented:
but please, finish those beers first before i get 10 more emails of you guys shouting at each other, forcing me to hit the unsubscribe button again ;-)

ksooo commented:
Using value 0 sometimes for a valid index, sometimes for an invalid index cannot really work. Defining a constant for the one of the two zeros makes things not better, but it makes the code more errorprone. How should I know that I must NOT compare a given index against the constant "INVALID_INDEX" if I want to know whether the value is an invalid index?

Obviously we are in a need of a unique value for an invalid client index.

From my point of, to fix this properly, we unfortunately need a PVR Addon API change. I see two approaches here:

1) Leave PVR_TIMER.iClientIndex type unsigned and define 0 as invalid index (including definition of a constant for this special value). Change type of CPVRTimerInfoTag.m_iClientIndex to unsigned, adapt Kodi init and comparing code to use the new constant. Adapt all addons no longer to use 0 index for valid timers and an API version bump must be done.

2) Change PVR.iClientIndex type to signed and define -1 as invalid index (again, including definition of a constant for that special value). Leave type of CPVRTimerInfoTag.m_iClientIndex signed, adapt Kodi init and comparing code to use the new constant. No addon code changes needed (just the API version bump).

Given the lower effort I'd prefer approach 2. What do people think?

janbar commented:
Today index '0' is reserved for PVR_TIMER_NO_PARENT. It is not usable as valid timer index because it cannot be a valid parent. I think pvr addon maintainers have fixed that by themselves by not using index 0.

janbar commented:
It seems assignments with magic number were useless because values were same as those initially set by CTOR. I removed these useless assignments and finally only CTOR assign a static value: I think we can avoid #define if Lars is in agreement.
Else i will add #define in header file PVRTimerInfoTag.h.

janbar committed:
[PVR] fix compilation warnings

Use definitely unsigned value for ClientIndexId as PVR API defines it.
Also it removes all useless assignment with magic number keeping initial
value from CTOR.

janbar commented:
Rebased again ;)

ksooo commented:
> Today index '0' is reserved for PVR_TIMER_NO_PARENT. It is not usable as valid timer index ...

Right. Overlooked this. This kicks my proposal number 2 out of the game. Then we're with proposal number 1.

So, what is left to do is 

* to add a constant to xbmc_pvr_types.h (not in PVRTimerInfoTag.h as the constant must be part of the PVR Addon API) like this:

<pre>
/*!
   * @brief special PVR_TIMER.iClientIndex value to indicate that a timer has no valid index.
   */
  const unsigned int PVR_TIMER_INVALID_CLIENT_INDEX = 0; /*!< @brief timer has no valid index.*/
</pre>

* to change Kodi code to use PVR_TIMER_INVALID_CLIENT_INDEX instead of 0 

* to review all PVR addons and to change code in case they use 0 as valid index today

In case at least one addon actually uses 0 index as valid value today (strictly speaking the following should also be done in case no addon actually uses index 0):

* to bump the addon api version in xbmc/addons/include/xbmc_pvr_types.h and addons/xbmc.pvr/addon.xml to 3.0.0 as this is a breaking change (addons using index 0 for their timers will break).

* to bump PVR Addon API version of all addons (no matter they used index 0 or not)

IMO, lots of work but this actually fixes the problem not having a unique value for an invalid client index. Anthing else is improvement, but no real fix.

janbar commented:
@ksooo here there is no API change. We just fix some warnings could make bugs.
And :) i just find a big bug caused by signed/unsigned conflict: ```349: if (!timerTag || timerTag->m_iClientIndex < 0)``` forbids deletion of my timers having index 2173071493.
Since commit 4bff347 , unsigned client indexes are cast as signed value making them less than zero if value are up to 2^31.
@fritsch @opdenkamp help !

ksooo commented:
@janbar I have to say that I'm clueless and don't know what else I can do, but obviously I'm not able to get over to you what I want to achieve. Of course there IS no API change in your PR right now, but I suggest we should do one to solve the actual problem instead of just making the compiler happy. 

Regarding the bug you have found: I'm not against making m_iClientIndex unsigned! I just trying to get over that from muy point of view your PR is not quite complete.

And now I'm done with this PR. This is eating up far too much of my spare time. Maybe I'm the bad guy now, but hope that the good guys you added in your last comment for help now may be with you.

hudokkow commented:
@ksooo, I think the real problem is the language barrier, not good guy/bad guy stuff. Problem exists and @janbar is trying to solve it. Maybe he didn't fully understand https://github.com/xbmc/xbmc/pull/7488#issuecomment-121597072

janbar commented:
Thanks hudo, things go ahead. According with ksooo, i break down this PR and make a new one to resolve side effect of signed value:  #7521

janbar commented:
I will close this PR and let who wants solve signed/unsigned conflict.
